File: Protogen/loxim_protocol/java/src/main/java/pl/edu/mimuw/loxim/protocol/packages_data/DatetimePackage.java
package pl.edu.mimuw.loxim.protocol.packages_data;

import java.util.EnumSet;
import pl.edu.mimuw.loxim.protogen.lang.java.template.ptools.Package;
import pl.edu.mimuw.loxim.protogen.lang.java.template.exception.ProtocolException;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageInputStreamReader;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageOutputStreamWriter;


import pl.edu.mimuw.loxim.protocol.enums.*;


public  class DatetimePackage extends Package {
		public final static long ID=(long)12;
	
    	/**
**/
    	private DatePackage date=null;
    	/**
**/
    	private TimePackage time=null;
    	
	public DatetimePackage()
	{
	}

		public DatetimePackage(
		    	DatePackage a_date,
	        	TimePackage a_time    	)
	{
        	date=a_date;
        	time=a_time;
    
	}
		
		@Override
	public long getPackageType() {
		return (long)ID;
	}
		
	@Override
	protected void deserializeW(PackageInputStreamReader reader)
			throws ProtocolException {
		super.deserializeW(reader);
		
				try{
	    			date=new DatePackage();
date.deserializeContent(reader);
					time=new TimePackage();
time.deserializeContent(reader);
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}
			}
	
	@Override
	protected void serializeW(PackageOutputStreamWriter writer)
			throws ProtocolException {
		super.serializeW(writer);
				try{
	    			date.serializeContent(writer);
		
					time.serializeContent(writer);
		
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}	
			}
	
	@Override
	public boolean equals(Object obj) {
		if(!super.equals(obj))
			return false;
		DatetimePackage p=(DatetimePackage)obj;
	    if (!(((date==p.getDate())||((date!=null)&&(p.getDate()!=null)&&((date).equals(p.getDate())))))) 
return false;
		if (!(((time==p.getTime())||((time!=null)&&(p.getTime()!=null)&&((time).equals(p.getTime())))))) 
return false;
				return true;
	}
	
//================= GETTERS AND SETTERS ===================

        	public DatePackage getDate(){return date;};
		
    	public void setDate(DatePackage a_date){ date=a_date;};
		
        	public TimePackage getTime(){return time;};
		
    	public void setTime(TimePackage a_time){ time=a_time;};
		
    
	 
}

File: Protogen/loxim_protocol/cpp/protocol/packages_data/DatetimePackage.cpp
#include <string.h>
#include <stdlib.h>
#include "DatetimePackage.h"
#include "ptools/PackageBufferWriter.h"
#include "ptools/PackageBufferReader.h"

using namespace protocol;

DatetimePackage::DatetimePackage(\
		DatePackage* a_date,\
		TimePackage* a_time
		)
{
	date=a_date;
	time=a_time;
}

DatetimePackage::DatetimePackage()
{
	date=NULL;
	time=NULL;
}

DatetimePackage::~DatetimePackage()
{
	if(date) delete date;
	date=NULL;
	if(time) delete time;
	time=NULL;
}

void DatetimePackage::serializeW(PackageBufferWriter *writer)
{
	Package::serializeW(writer);
	date->serializeContent(writer);
	time->serializeContent(writer);
}

bool DatetimePackage::deserializeR(PackageBufferReader *reader)
{
	if (!Package::deserializeR(reader)) return false;
	date=new DatePackage();
	if (!date->deserializeContent(reader)) return false;
	time=new TimePackage();
	if (!time->deserializeContent(reader)) return false;
	return true;
}

bool DatetimePackage::equals(Package *_package)
{
	if (!_package) return false;
	if (_package->getPackageType()!=this->getPackageType()) return false;
	if (!Package::equals(_package)) return false;
	DatetimePackage* p=(DatetimePackage*)_package;
	if (!((!getDate() && ! p->getDate())||(getDate() && p->getDate() && (getDate()->equals(p->getDate()))))) return false;
	if (!((!getTime() && ! p->getTime())||(getTime() && p->getTime() && (getTime()->equals(p->getTime()))))) return false;
	return true;
}
